core/src/autogluon/core/pseudolabeling/pseudolabeling.py

```diff
@@ -0,0 +1,218 @@
+
+    if num_models < 2:
+        raise Exception('Ensemble pseudo labeling was enabled, but only one model was trained')
+
+    if num_models != num_models:
+        logging.warning(f'Ensemble pseudo labeling expected {original_k}, but only {num_models} fit.')
+
+    if predictor.problem_type in PROBLEM_TYPES_CLASSIFICATION:
+        return filter_ensemble_classification(predictor=predictor, unlabeled_data=unlabeled_data,
+                                              leaderboard=leaderboard, num_models=num_models)
+    else:
+        return filter_pseudo_std_regression(predictor=predictor, unlabeled_data=unlabeled_data,
+                                            leaderboard=leaderboard, num_models=num_models)
+
+
+def filter_pseudo_std_regression(predictor, unlabeled_data: pd.DataFrame, num_models, leaderboard,
+                                 lower_bound: float = -0.25, upper_bound: float = 0.25):
+    """
+    Predicts on unlabeled_data using the top k models. Then gets standard deviation of each
+    row's prediction and the standard deviation across all rows of standard deviations and
+    incorporates rows with a z-score in the range of lower_bound and upper_bound.
+
+    Parameters:
+    -----------
+    predictor: Fitted tabular predictor that ensembles multiple models
+    unlabeled_data: Unlabeled data for top k models to predict on
+    leaderboard: pd.DataFrame of leaderboard of models in AutoGluon based on validation score
+    num_models: Number of top models to ensemble
+    lower_bound: Lower threshold that z-score needs to exceed in order to
+        incorporate
+    upper_bound: Upper threshold that z-score needs to be below to incorporate
+
+    Returns:
+    --------
+    pd.Series of indices that met pseudo labeling requirements
+    """
+    top_k_models_list = leaderboard.head(num_models)['model']
+    top_k_preds = None
+    best_model_preds = None
+
+    for model in top_k_models_list:
+        y_test_pred = predictor.predict(data=unlabeled_data, model=model)
+
+        if best_model_preds is None:
+            best_model_preds = y_test_pred
+
+        if model == top_k_models_list[0]:
+            top_k_preds = y_test_pred
+        else:
+            top_k_preds = pd.concat([top_k_preds, y_test_pred], axis=1)
+
+    top_k_preds = top_k_preds.to_numpy()
+    preds_sd = pd.Series(data=np.std(top_k_preds, axis=1), index=unlabeled_data.index)
+    preds_z_score = (preds_sd - preds_sd.mean()) / preds_sd.std()
+    df_filtered = preds_z_score.between(lower_bound, upper_bound)
+
+    return df_filtered[df_filtered], y_test_pred
+
+
+def filter_ensemble_classification(predictor, unlabeled_data: pd.DataFrame, leaderboard,
+                                   num_models, threshold: float = 0.95):
+    """
+    Calculates predictive prob of unlabeled data by predicting with top k models then averages
+    and selects rows where max predictive prob is above threshold. Then samples minimum
+    bin count from all bins.
```

Comment:
description should again be made more clear (and don't use abbreviations in descriptions). Eg: explain average is taken per-row over predictions from top k models. And add "where confidence exceeds threshold (with confidence defined as the predicted probability of the most likely class)".  The final sentence should also be made much more explicit re what is happening (eg. defining what are "bins").

Reply:
"Calculates predictive probabilities of unlabeled data by predicting with top k models
    then averages pre-row over predictions from top k models and selects rows where confidence
    (predicted probability of the most likely class) is above threshold. Then samples minimum
    bin count from all bins, where rows of averaged predictions are binned into peak predicted
    probability class."